OpenRA.Game/GameRules/RulesetWatcher.cs

```diff
@@ -0,0 +1,162 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Timers;
+using OpenRA.FileSystem;
+
+namespace OpenRA.GameRules
+{
+	public class RulesetWatcher : IDisposable
+	{
+		readonly World world;
+		readonly ModData modData;
+		readonly Dictionary<string, string> watchFiles = new();
+		readonly Timer timer;
+		readonly HashSet<string> fileQueue = new();
+		readonly FileSystemWatcher watcher;
+
+		volatile bool isEnabled;
+		bool isDisposed;
+
+		public bool IsEnabled => isEnabled;
+
+		public RulesetWatcher(World world, ModData modData)
+		{
+			this.world = world;
+			this.modData = modData;
+
+			var files = modData.Manifest.Rules
+				.Concat(modData.Manifest.Weapons)
+				.Concat(modData.Manifest.Sequences);
+			foreach (var file in files)
+			{
+				string filename;
+				using (var fs = modData.DefaultFileSystem.Open(file) as FileStream)
+					filename = fs?.Name;
+
+				if (string.IsNullOrEmpty(filename))
+					continue;
+
+				var fullPath = Path.GetFullPath(filename);
+				watchFiles[fullPath] = file;
+			}
+
+			timer = new Timer(10) { AutoReset = false };
+			timer.Elapsed += RefreshTimer;
+
+			watcher = new FileSystemWatcher(modData.Manifest.Package.Name)
+			{
+				IncludeSubdirectories = true
+			};
+			watcher.Changed += FileChanged;
+
+			foreach (var file in watchFiles.Keys.Distinct())
+			{
+				watcher.Filters.Add(Path.GetFileName(file));
+			}
+		}
+
+		public void StartWatching()
+		{
+			if (isDisposed)
+				throw new ObjectDisposedException(nameof(RulesetWatcher));
+
+			watcher.EnableRaisingEvents = true;
+			isEnabled = true;
+			RestartTimerIfEnabled();
+		}
+
+		public void StopWatching()
+		{
+			if (isDisposed)
+				throw new ObjectDisposedException(nameof(RulesetWatcher));
+
+			isEnabled = false;
+			timer.Stop();
+			watcher.EnableRaisingEvents = false;
+
+			lock (fileQueue)
+				fileQueue.Clear();
+		}
```

Comment:
I think a single method makes sense here, since then you can do the `isDisposed` check once and reduce a bit of code duplication. It is also easier to use by external classes since they don't need to choose between two methods.

```cs
public void ToggleWatching(bool toggleValue)
{
    if (isDisposed)
    	throw new ObjectDisposedException(nameof(RulesetWatcher));
    	
    watcher.EnableRaisingEvents = toggleValue
    isEnabled = toggleValue
    
    if (toggleValue)
        RestartTimerIfEnabled();
    else
    {
        timer.stop();
        lock (fileQueue)
				fileQueue.Clear();
	}
}